<a href="https://colab.research.google.com/github/ChristianKitte/HelloKNN/blob/main/KNN_sklearn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Die eigentliche Lösung der Aufgabe ist ganz am Ende zu finden...**

In [11]:
#https://www.python-course.eu/k_nearest_neighbor_sklearn.php

Im ersten Schritt importiere ich alle im folgenden benötigten Bibliotheken

In [12]:
%matplotlib inline

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn import datasets
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

print("Bibliotheken importiert....")

Bibliotheken importiert....


Um den Code übersichtlicher zu halten, lagere ich die Funktionalität für die Normalisierung in eine eigene Funktion aus. Ich verwende hierbei den Algorithmus von sklearn mit der Normierungsmethode "max". Es scheint, dass dieser im einfachsten Fall einfach alles durch den maximalen Wert teilt. Hierdurch ist der größte Wert 1, der kleinste geht gegen 0.

In [13]:
# https://scikit-learn.org/stable/modules/preprocessing.html
# sklearn teilt die Werte enfach durch max, wie es hier scheint (zumindest bei den verwendeten Paramter "max")

# https://scikit-learn.org/stable/auto_examples/preprocessing/plot_all_scaling.html#sphx-glr-auto-examples-preprocessing-plot-all-scaling-py
# https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler

from sklearn.preprocessing import MinMaxScaler

scaler=0

def normalizeDataSklearn(values):
  # normalized_array = preprocessing.normalize(values, axis=0,norm="max")
  
  global scaler
  scaler = MinMaxScaler()
  scaler.fit(values)
  
  normalized_array = scaler.transform(values)

  #print(normalized_array)
  return normalized_array

print("Funktion normalizeDataSklearn wurde erstellt....")

Funktion normalizeDataSklearn wurde erstellt....


Als nächstes teile ich das dataset in Test- und Trainingsdaten. Dies habe ich in der vorhergehenden Übung (Regression) noch nicht gemacht. Es scheint mir aber mittlerweile auch bei den Übungen sinnvoll, da nebenbei auch das dataset durchgemischt wird. (Auch wenn es bei KNN nicht notwendig scheint, da gegen alle vorhandenen Elemente gerechnet wird)

Ich erstelle hierbei zwei Versionen. Eine ohne vorhergehende Normalisierung, eine mit. Bei KNN ist die Normalisierung vor allem bei großen Differenzen innerhalb der Zahlenbereich notwendig, um ein shadowing vorzubeugen.

Bei dem IRIS Datensatz halte ich dies aber für noch nicht notwendig.

In [14]:
# https://scikit-learn.org/stable/auto_examples/datasets/plot_iris_dataset.html
# 'feature_names': ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)'],
# 'target_names': ['setosa', 'versicolor', 'virginica'] mit 0, 1, 2 

# https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html
# Split arrays or matrices into random train and test subsets ==> Es erfolgt ein Shuffle, daher alleine 
# schon deswegen sinnvoll! randdom_state ermöglicht Replizierbarkeit. Rückgabe ist ein Array mit den
# Arrays.

ds = datasets.load_iris()
data=ds.data
labels=ds.target

dataNorm = normalizeDataSklearn(data)
resArrayNorm = train_test_split(dataNorm, labels, train_size=0.8, test_size=0.2, random_state=1)
trainDataNorm, testDataNorm, trainLabels, testLabels = resArrayNorm 
#print(dataNorm)
#print(data)

resArray = train_test_split(data, labels, train_size=0.8, test_size=0.2, random_state=1)
trainData, testData, trainLabels, testLabels = resArray 

#ds
#labels
#target

#print(trainData)
#print(trainLabels)

print("Daten vorhanden und vorbereitet")

Daten vorhanden und vorbereitet


Nachdem alle Daten vorbereitet sind, trainiere ich analog zu oben zwei Klassifizierer (einmal mit normalisierten Daten, einmal ohne)

In [15]:
knnClassifier = KNeighborsClassifier(algorithm="auto",
                                     metric='euclidean',
                                     n_neighbors=8)
knnClassifier.fit(trainData, trainLabels)

knnClassifierNorm = KNeighborsClassifier(algorithm="auto",
                                     metric='euclidean',
                                     n_neighbors=8)
knnClassifierNorm.fit(trainDataNorm, trainLabels) 

print("Klassifizierer ist vorbereitet and antrainiert")

Klassifizierer ist vorbereitet and antrainiert


Abschließend führe ich mit beiden Klassifizierer Vorhersagen aus. Einmal über die Testdaten, einmal über die Trainingsdaten und einmal über alle Daten.

In [16]:
# https://scikit-learn.org/stable/modules/model_evaluation.html
# https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html#sklearn.metrics.accuracy_score
# "...the set of labels predicted for a sample must exactly match..."

predLabelsTest = knnClassifier.predict(testData)
predLabelsTrain = knnClassifier.predict(trainData)
predLabelsAll = knnClassifier.predict(data)

predLabelsTestNorm = knnClassifierNorm.predict(scaler.transform(testData))
predLabelsTrainNorm = knnClassifierNorm.predict(scaler.transform(trainData))
predLabelsAllNorm = knnClassifierNorm.predict(scaler.transform(data))

#print("Vorhersage:")
#print(predLabelsTest)
#print(predLabelsTrain)
#print(predLabelsAll)

#print("Originalwerte:")
#print(testLabels)
#print(trainLabels)
#print(labels)

print()
print("Accuracy Testdaten: ", accuracy_score(predLabelsTest, testLabels))
print("Accuracy Trainingsdaten: ", accuracy_score(predLabelsTrain, trainLabels))
print("Accuracy alle Daten: ", accuracy_score(predLabelsAll, labels))

print()
print("Accuracy (mit Normierung) Testdaten: ", accuracy_score(predLabelsTestNorm, testLabels))
print("Accuracy (mit Normierung) Trainingsdaten: ", accuracy_score(predLabelsTrainNorm, trainLabels))
print("Accuracy (mit Normierung) alle Daten: ", accuracy_score(predLabelsAllNorm, labels))

print()
print("Fertig...")


Accuracy Testdaten:  1.0
Accuracy Trainingsdaten:  0.9833333333333333
Accuracy alle Daten:  0.9866666666666667

Accuracy (mit Normierung) Testdaten:  1.0
Accuracy (mit Normierung) Trainingsdaten:  0.9666666666666667
Accuracy (mit Normierung) alle Daten:  0.9733333333333334

Fertig...


Die Ergebniss mit und ohne Normierung unterscheiden sich ein wenig, wobei die ohne etwas besser sind:

* Accuracy Testdaten:  1.0
* Accuracy Trainingsdaten:  0.9833333333333333
* Accuracy alle Daten:  0.9866666666666667
---

* Accuracy (mit Normierung) Testdaten:  1.0
* Accuracy (mit Normierung) Trainingsdaten:  0.9666666666666667
* Accuracy (mit Normierung) alle Daten:  0.9733333333333334

Dies bestätigt meine oben gemachte Vermutung.

Nach Abschluss der Vorbereitungen (und einigen Experimenten) komme ich nun zur Lösung der zweiten Aufgabe der 10. EA - Instanzbasiertes Lernen:

**Aufgabe:**

In the logistic regression example, I gave you a new iris data:
4.8,2.5,5.3,2.4

Please classify this flower using KNN.


In [17]:
values=np.array([4.8,2.5,5.3,2.4])

pred = knnClassifier.predict(values.reshape(-1,4))
predNorm = knnClassifier.predict(values.reshape(-1,4))

#'target_names': ['setosa', 'versicolor', 'virginica'] mit 0, 1, 2 

print(pred)
print(predNorm)

[2]
[2]


**Lösung** 

Beide Klassifizierer sagen für das unbekannt Element vorher, dass es sich um eine Iris Virginica handelt. Hiermit wird auch das Ergebnis aus der Übung zur Regression bestätigt.